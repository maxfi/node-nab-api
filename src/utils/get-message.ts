import { isPlainObject } from 'is-plain-object';

const convert = (x: object) => {
  if (isPlainObject(x)) {
    return Object.entries(x).reduce((acc, value) => {
      const [k, v] = value;
      acc[k] = convert(v);
      return acc;
    }, {} as Record<string, unknown>);
  }

  return { _text: x };
};

export const getMessage = (x: object) => {
  if (!isPlainObject(x)) throw new Error('Requires a plain object');
  return convert(x);
};
