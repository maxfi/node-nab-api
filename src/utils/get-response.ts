import { isPlainObject } from 'is-plain-object';

const convert = (x: any) => {
  if (!isPlainObject(x)) return x;
  if (x._text) return x._text;
  return Object.entries(x).reduce((acc, value) => {
    const [k, v] = value;
    acc[k] = convert(v);
    return acc;
  }, {} as Record<string, string>);
};

export const getResponse = (x: object) => {
  if (!isPlainObject(x)) throw new Error('Requires a plain object');
  return convert(x);
};
