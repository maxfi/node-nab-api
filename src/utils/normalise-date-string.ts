import parse from 'date-fns/parse';

const pad = (str: string) => (str.length === 1 ? '0' + str : str);

export const normaliseDateString = (str: string) => {
  const [d, o] = str.split('000+');
  const offsetMins = parseInt(o, 10);
  const hours = pad(Math.floor(offsetMins / 60).toString());
  const mins = pad((offsetMins % 60).toString());
  const fixed = `${d}+${hours}:${mins}`;
  return parse(fixed, 'yyyyddMMHHmmssSSSXXX', new Date()).toISOString();
};
