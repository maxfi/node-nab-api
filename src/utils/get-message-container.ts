import { getMessage } from './get-message';

export const getMessageContainer = (input: {
  messageId: string;
  credentials: { merchantId: string; password: string };
  requestType: string;
  dataElementName: string;
  payload: object;
  options: { timeout?: string };
}) => {
  input.options = input.options || {};
  return {
    _declaration: {
      _attributes: {
        version: '1.0',
        encoding: 'UTF-8',
      },
    },
    NABTransactMessage: {
      MessageInfo: {
        messageID: {
          _text: input.messageId,
        },
        messageTimestamp: {
          _text: new Date().toISOString(),
        },
        timeoutValue: {
          _text: input.options.timeout || '60',
        },
        apiVersion: {
          _text: 'spxml-3.0',
        },
      },
      MerchantInfo: {
        merchantID: {
          _text: input.credentials.merchantId,
        },
        password: {
          _text: input.credentials.password,
        },
      },
      RequestType: {
        _text: input.requestType,
      },
      [input.dataElementName]: {
        [input.dataElementName + 'List']: {
          _attributes: {
            count: '1',
          },
          [input.dataElementName + 'Item']: Object.assign(
            { _attributes: { ID: '1' } },
            getMessage(input.payload)
          ),
        },
      },
    },
  };
};
