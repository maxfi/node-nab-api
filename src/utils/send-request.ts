import fetch from 'node-fetch';
import { js2xml, xml2js } from 'xml-js';
import { getResponse } from './get-response';

export const toXml = (x: object) => js2xml(x, { spaces: 4, compact: true });
export const fromXml = (str: string) => xml2js(str, { compact: true });

/**
 * Make the request to the NAB payment gateway API.
 * Converts the payload to XML and converts the response to Object.
 * @param  url - Endpoint URL.
 * @param  payload - Request body.
 * @return Returns the API response as Object.
 */
export const sendRequest = async (url: string, payload: object) => {
  const xml = toXml(payload);
  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'text/xml' },
    body: xml,
  });
  if (!response.ok) {
    const { status, statusText } = response;
    console.log('REQUEST:', url, ':', xml);
    throw new Error(`Bad response from NAB API: ${status} - ${statusText}`);
  }
  const body = await response.text();
  return getResponse(fromXml(body));
};
