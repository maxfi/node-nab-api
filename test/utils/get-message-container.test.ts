import { getMessageContainer } from '../../src/utils/get-message-container';
import { toXml } from '../../src/utils/send-request';
import { isoDateStringRegex } from '../helpers/iso-date-string-regex';

const payload = {
  messageId: '8af793f9af34bea0ecd7eff711c9d3',
  credentials: {
    merchantId: 'XYZ00',
    password: 'abcd1234',
  },
  requestType: 'addToken',
  dataElementName: 'Token',
  payload: {
    cardNumber: '4444333322221111',
    expiryDate: '11/15',
    tokenType: '1',
  },
  options: {},
};

const messageTimeStampPropertyMatcher = {
  NABTransactMessage: {
    MessageInfo: {
      messageTimestamp: { _text: expect.stringMatching(isoDateStringRegex) },
    },
  },
};

const removeTimestamp = (x: string) =>
  x.replace(/<messageTimestamp>.*<\/messageTimestamp>/, '');

test('getMessageContainer', () => {
  const result = getMessageContainer(payload);
  expect(result).toMatchSnapshot(messageTimeStampPropertyMatcher);
});

test('getMessageContainer should create the correct object to be parsed to XML', async () => {
  const xml = removeTimestamp(toXml(getMessageContainer(payload)));
  expect(xml).toMatchSnapshot();
});
