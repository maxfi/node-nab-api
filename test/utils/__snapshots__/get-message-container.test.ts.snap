// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMessageContainer 1`] = `
Object {
  "NABTransactMessage": Object {
    "MerchantInfo": Object {
      "merchantID": Object {
        "_text": "XYZ00",
      },
      "password": Object {
        "_text": "abcd1234",
      },
    },
    "MessageInfo": Object {
      "apiVersion": Object {
        "_text": "spxml-3.0",
      },
      "messageID": Object {
        "_text": "8af793f9af34bea0ecd7eff711c9d3",
      },
      "messageTimestamp": Object {
        "_text": StringMatching /\\^\\\\d\\{4\\}-\\[01\\]\\\\d-\\[0-3\\]\\\\dT\\[0-2\\]\\\\d:\\[0-5\\]\\\\d:\\[0-5\\]\\\\d\\\\\\.\\\\d\\+\\(\\[\\+-\\]\\[0-2\\]\\\\d:\\[0-5\\]\\\\d\\|Z\\)\\$/u,
      },
      "timeoutValue": Object {
        "_text": "60",
      },
    },
    "RequestType": Object {
      "_text": "addToken",
    },
    "Token": Object {
      "TokenList": Object {
        "TokenItem": Object {
          "_attributes": Object {
            "ID": "1",
          },
          "cardNumber": Object {
            "_text": "4444333322221111",
          },
          "expiryDate": Object {
            "_text": "11/15",
          },
          "tokenType": Object {
            "_text": "1",
          },
        },
        "_attributes": Object {
          "count": "1",
        },
      },
    },
  },
  "_declaration": Object {
    "_attributes": Object {
      "encoding": "UTF-8",
      "version": "1.0",
    },
  },
}
`;

exports[`getMessageContainer should create the correct object to be parsed to XML 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<NABTransactMessage>
    <MessageInfo>
        <messageID>8af793f9af34bea0ecd7eff711c9d3</messageID>
        
        <timeoutValue>60</timeoutValue>
        <apiVersion>spxml-3.0</apiVersion>
    </MessageInfo>
    <MerchantInfo>
        <merchantID>XYZ00</merchantID>
        <password>abcd1234</password>
    </MerchantInfo>
    <RequestType>addToken</RequestType>
    <Token>
        <TokenList count=\\"1\\">
            <TokenItem ID=\\"1\\">
                <cardNumber>4444333322221111</cardNumber>
                <expiryDate>11/15</expiryDate>
                <tokenType>1</tokenType>
            </TokenItem>
        </TokenList>
    </Token>
</NABTransactMessage>"
`;
